### PR TITLE
[Common] Fused cast transpose kernels refactoring

### PR DESCRIPTION
**Pull request by Oleg-Goncharov**
## Description
Existing code of the fused cast transpose kernels is replicated for different scenarios (i.e. +dbias, +dactivation) with only small specific modifications. Replacing it with a single function template makes the code easier to support and allows adding new features in a simple way (e.g., scaling, JIT compilation).

## Changes:
- Moved all the gated activation functions into a separate file `gated_act_cast_transpose.cu`
- Created a single function (and kernel) template for dbias and dgelu
- Changed the value of the `n_warps_per_tile` parameter from 4 to 8, which slightly improves the performance on the H100 HBM3

The following table provides the runtime of the previous and the new version of the fused cast transpose kernels on the H100 HBM3. The new version is benchmarked for two values of the `n_warps_per_tile` parameter (4 and 8):

Cast_transpose_fused-Template Code

### CODE REVIEW

File: transformer_engine/common/transpose/cast_transpose_fusion.cu [hunk 1]
```diff
@@ -134,1451 +119,596 @@ struct CTDBiasDGeluParam {
...
+
+            if constexpr (IS_DBIAS) {
+                if (workspace->data.dptr == nullptr) {
+                    populate_cast_transpose_dbias_workspace_config(*cast_output,
+                                                                   workspace, nvec_out);
+                    return;
+                }
+            }
+
+            CheckInputTensor(input, "cast_transpose_dbias_dact_input");
```
Comment (timmoon10) [inline transformer_engine/common/transpose/cast_transpose_fusion.cu:640]:
```suggestion
            CheckInputTensor(input, "cast_transpose_fused_input");
```

File: transformer_engine/common/transpose/cast_transpose_fusion.cu [hunk 2]
```diff
@@ -134,1451 +119,596 @@ struct CTDBiasDGeluParam {
...
+
+    const size_t stride = row_length / nvec_in;
+    const size_t output_stride = num_rows / nvec_out;
+    size_t current_stride = warp_id_in_tile * n_iterations * nvec_out * stride;
+    unsigned int my_place = (my_id_in_warp + THREADS_PER_WARP - warp_id_in_tile * n_iterations) %
+                             THREADS_PER_WARP;
+    CType max = 0;
+    const CType scale = param.scale_ptr != nullptr ? *param.scale_ptr : 1;
+
+    if constexpr (IS_DBIAS) {
```
[the remark this answers is not in the document]
Comment (Oleg-Goncharov) [reply transformer_engine/common/transpose/cast_transpose_fusion.cu:269]: Thanks a lot for the note! A few extra will come with the scaling PR

**Comment by phu0ngng** [comment]
Hi, I think templating DACT and DBIAS is a good idea. Great work!

I have little concern as now we have `gated_act_cast_transpose.cu`  but `act_cast_transpose` is still secretly in `cast_transpose.cu`. I think it is better to either keep all related cast transpose functions in one place OR split both `gated_act` and `act` related functions into two additional files.

**Comment by Oleg-Goncharov** [comment]
Agree, to have it consistent across other cast-transpose files, I reverted the split. There is a single file `cast_transpose_fusion.cu` as before.

**Review by phu0ngng — APPROVED** [review]

**Review by timmoon10 — APPROVED** [review]

**Pull request merged**